packages/slice-machine/src/features/autoSave/useAutoSave.tsx

```diff
@@ -0,0 +1,114 @@
+import { useCallback, useState, useEffect } from "react";
+import { uniqueId } from "lodash";
+import { toast } from "react-toastify";
+import { Box, Button, Text } from "@prismicio/editor-ui";
+
+type UseAutoSaveArgs = {
+  debounceDelay?: number;
+  errorMessage?: string;
+  retryDelay?: number;
+  retryMessage?: string;
+};
+
+type AutoSaveState = {
+  pendingSave: (() => Promise<unknown>) | null;
+  nextSave: (() => Promise<unknown>) | null;
+};
+
+export type AutoSaveStatusType = "saving" | "saved" | "error";
+
+export const useAutoSave = (args: UseAutoSaveArgs = {}) => {
+  const {
+    debounceDelay = 500,
+    errorMessage = "An error happened while saving",
+    retryDelay = 2000,
+    retryMessage = "Retry",
+  } = args;
+  const [autoSaveState, setAutoSaveState] = useState<AutoSaveState>({
+    pendingSave: null,
+    nextSave: null,
+  });
+  const [autoSaveStatusInternal, setAutoSaveStatusInternal] =
```

Comment:
Hum not sure to understand the proposed naming too. Mostly because there is a diff between "state" and "status" in the naming

It was a bit hard here for the naming.
So the goal is really differentiate the state that is purely internal with a debounce so autoSave hook to transmit that a save is in progress if the success is superfast.

It means we need 2 state for the status. One internal and one share externally. Maybe I could also try to rewrite it with only one state but 2 properties, WDYT?

Something like this (inlined here):

```ts
  const [autoSaveStack, setAutoSaveStack] = useState<AutoSaveStack>({
    pendingSave: null,
    nextSave: null,
  });
  const [autoSaveStatus, setAutoSaveStatus] = useState<{
    debounced: AutoSaveStatusType;
    immediate: AutoSaveStatusType;
  }>({
    debounced: "saved",
    immediate: "saved", // or "undebounced" or ??
  });
```

With that it's maybe clear what is the purpose of each useState and reduces the overall complexity.

Reply:
Let's go with `actual` thanks

Reply:
So I did it differently at the end. With the new requirement from Benjamin that want the opposite of a status that only show if it takes longer than x ms. The goal is to have a status that always update to display the loading state but will take 300ms longer to remove that it's saved or failed. So that we can always smoothly see that something happened.

With that said, at the end it was not better to mix both state as we really want to have two lifecycles. So I have one status delayed and one actual. I think it's clearer now.

I let you check 🙂